source/sonata_simulation.rst

```diff
@@ -96,6 +97,27 @@ example::
        "integration_method" : 2
   }
 
+Format of the electrodes_file
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The format of the weights file is described below:
+
+.. table::
+
+  ============================= ======================= ========== =============== ============= ==========================================
+  Group                         Field                   Type       Shape           Requirement   Description
+  ============================= ======================= ========== =============== ============= ==========================================
+  /electrodes/{electrodename}   layer                   utf8       1               Optional      Layer of the circuit in which {electrodename} is located. If the electrode is in a region without cortical layers, then "NA". If the electrode is outside the brain, then "Outside"
+  /electrodes/{electrodename}   position                float32    3               Mandatory      Position of {electrodename} in microns, in cartesian coordinates
+  /electrodes/{electrodename}   region                  utf8       1               Optional      Region in which {electrodename} is located
+  /electrodes/{electrodename}   type                    utf8       1               Optional      Either EEG or LFP
+  /electrodes/{electrodename}   offset                  uint64     1               Optional      Column index of the coefficients for {electrodename} in /electrodes/{population_name}/{node_id}
+  /electrodes/{population_name} {node_id}               float64    N_seg x N_elec  Mandatory     Scaling factor for each segment in the corresponding neuron, in mV/nA
+  /{population_name}            node_ids                uint64     N_nodes         Mandatory     List of node ids
+  /{population_name}/offsets    {node_id}               uint64     N_sec           Mandatory     For each section, number of entries in /electrodes/{population_name}/{node_id} between start of previous section and start of current section. Equivalent to number of segments in previous section
+  /{population_name}/sec_ids    {node_id}               unit64     N_seg           Mandatory     For each segment, index of the section to which it belongs
```

Comment:
I'm concerned about the layout of this portion: it appears that there are datasets for potentially millions of cells:

```
/nodeA                   Group
/nodeA/node_ids          Dataset {2}
/nodeA/offsets           Group
/nodeA/offsets/0         Dataset {330}
(...potentially millions of datasets...)
/nodeA/offsets/N         Dataset {245}

/nodeA/sec_ids           Group
/nodeA/sec_ids/0         Dataset {1644}
(...potentially millions of datasets...)
/nodeA/sec_ids/N         Dataset {1243}
```

My recollection is that having this order of magnitude is not great for performance, and can result in significant slowdowns.  I believe that HPC have looked at this, in the context of morphology containers: perhaps @pramodk / @1uc / @matz-e remember better than I do, and can comment.

Reply:
After conducting tests and discussing with @1uc  @matz-e  @joseph-tharayil , we observed that the creation of the HDF5 file slows down significantly after approximately 1 million datasets. This issue has been discussed in detail in the following GitHub threads: [h5py issue #1055](https://github.com/h5py/h5py/issues/1055) and [Hierarchical-Localization pull request #193](https://github.com/cvg/Hierarchical-Localization/pull/193).

Considering our situation with potentially several million node_ids, the previously mentioned solution may not be sustainable. As an alternative, I propose a structure similar to reports, utilizing the following layout:
```
/electrodes/nodeA/data    Dataset {total_sections, nelectrodes}
/nodeA/sec_ids            Dataset {total_sections}
/nodeA/node_ids           Dataset {n_node_ids}
/nodeA/offsets            Dataset {n_node_ids+1}
```
To obtain data from a node_id with index i, we can access the data dataset as follows:
`subset = electrodes_dataset[offsets_dataset[i]:offsets_dataset[i+1], :]`

I have personally tested this approach with approximately 5 million node_ids, and the performance for both writing and reading the file was satisfactory.

Let me know if you have any further questions or concerns.

Reply:
Can you update the sample file, please?

Reply:
Done!